exareme2/controller/services/exaflow/execution_engine.py

```diff
@@ -0,0 +1,43 @@
+from typing import List
+
+from exareme2.controller import logger as ctrl_logger
+from exareme2.controller.services.exaflow.tasks_handler import ExaFlowTasksHandler
+
+
+class AlgorithmExecutionEngine:
+    """
+    The AlgorithmExecutionEngine is the class used by the algorithms to communicate with
+    the workers of the system. An AlgorithmExecutionEngine object is passed to all algorithms
+    """
+
+    def __init__(
+        self,
+        request_id: str,
+        context_id: str,
+        csv_paths_per_worker_id: dict,
+        tasks_handlers: List[ExaFlowTasksHandler],
+    ):
+        self._logger = ctrl_logger.get_request_logger(request_id=request_id)
+        self._context_id = context_id
+        self._tasks_handlers = tasks_handlers
+        self._csv_paths_per_worker_id = csv_paths_per_worker_id
+
+    def run_algorithm_step(
+        self,
+        func,
+        positional_args,
+    ) -> List[dict]:
+        results = []
+        for task_handler in self._tasks_handlers:
+            current_positional_args = positional_args
+            current_positional_args["csv_paths"] = self._csv_paths_per_worker_id[
+                task_handler.worker_id
+            ]
+            results.append(
+                task_handler.run_step(
+                    method_name=func,
+                    params=current_positional_args,
+                )
+            )
```

Comment:
This should be executed in parallel, otherwise it will never finish when the aggregation server is added.